Image: Implement 16-bit unorm and uint formats

BlueCube3310 opened the pull request:
Related to https://github.com/godotengine/godot-proposals/issues/7540 and https://github.com/godotengine/godot-proposals/issues/7914

Adds support for 16-bit `unorm` and `uint` formats to `Image`. This will allow storing certain data more efficiently, such as terrain heightmaps. 

Since `unorm` formats aren't explicitly required by Vulkan, as a fallback they will get converted to `float` in order to keep the bit depth. That said, the difference in supported devices is ~5% (According to https://vulkan.gpuinfo.org/listlineartilingformats.php)

Why both `uint16` and `unorm16`?
`unorm16` isn't as widely supported as `uint16`. As a result, many devices (mobile and web) would need to fall back to `float32`. This should be as seamless as possible, but it makes the data twice as big, which works against what the user was likely trying to achieve. `uint16` makes it possible to store the same data in the same amount of bits.

Why not use `uint16` only, then?
`uint16` (as well as other uint formats) are handled in a slightly different way than unorm. Notably, they have to be sampled using `usampler`, and only allow nearest-neighbor filtering when using the Compatibility renderer. `unorm16`, on the other hand, is sampled identically to `float` or `unorm8`.

My proposed solution would be to implement both formats and thoroughly document any caveats that may arise from either of them.

Testing project:
[16-bit-image-test.zip](https://github.com/user-attachments/files/20758786/16-bit-image-test.zip)

TODO:
- [x] Testing,
- [x] Fix bilinear resizing,
- [x] Documentation,
- [x] Image resizing,
- [x] Optimized conversion,
- [x] Mipmap generation,
- [x] DDS loading and saving,
- [x] KTX loading,
- [x] Channel detection

BlueCube3310 committed:
Image: Implement 16-bit unorm and uint formats

fire commented:
Will this cause noticable rendering problems with optimizations by struct  growth etc?

BlueCube3310 commented:
> Will this cause noticable rendering problems with optimizations by struct growth etc?

I don't fully understand the question, but it shouldn't cause issues, most of the new code is ran only when a texture is created.

BlueCube3310 commented:
Rebased on top of #106880, I've noticed it added casting from `DATA_FORMAT_D16_UNORM` to `Image::FORMAT_RH`. I'm not sure if it matters, but I changed it to `Image::FORMAT_R16` so that the `UNORM` type is kept.

BlueCube3310 commented:
CC @TokisanGames this may be of interest to you

TokisanGames commented:
Thanks for the ping. It shouldn't close  godotengine/godot-proposals#7540 without 32-bit support.

BlueCube3310 commented:
I think this is now ready for review/testing

Calinou commented on doc/classes/Image.xml:
```diff
@@ -760,7 +760,39 @@
-		<constant name="FORMAT_MAX" value="39" enum="Format">
+		<constant name="FORMAT_R16" value="39" enum="Format">
+			OpenGL texture format [code]GL_R16[/code] where there's one component, a 16-bit unsigned normalized integer value.
+			[b]Note:[/b] due to limited hardware support, it is mainly recommended to be used on desktop or console devices. It may be unsupported on mobile or web, and will consequently be converted to [constant FORMAT_RF].
```
```suggestion
			[b]Note:[/b] Due to limited hardware support, it is mainly recommended to be used on desktop or console devices. It may be unsupported on mobile or web, and will consequently be converted to [constant FORMAT_RF].
```

Same for the other notes.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

```gdscript
func _ready() -> void:
	var rgb16 := Image.create(1, 2, false, Image.FORMAT_RGB16)
	rgb16.set_pixel(0, 0, Color.RED)
	rgb16.set_pixel(0, 1, Color(100000, 1, 1))
	print(rgb16.get_pixel(0, 0))
	print(rgb16.get_pixel(0, 1))


	print()
	var rgb16i := Image.create(1, 2, false, Image.FORMAT_RGB16I)
	rgb16i.set_pixel(0, 0, Color.RED)
	rgb16i.set_pixel(0, 1, Color(100000, 1, 1))
	print(rgb16i.get_pixel(0, 0))
	print(rgb16i.get_pixel(0, 1))
```

Prints:

```
(1.0, 0.0, 0.0, 1.0)
(1.0, 1.0, 1.0, 1.0)

(1.0, 0.0, 0.0, 1.0)
(65535.0, 1.0, 1.0, 1.0)
```

Calinou commented on doc/classes/Image.xml:
```diff
@@ -760,7 +760,39 @@
 		</constant>
-		<constant name="FORMAT_MAX" value="39" enum="Format">
+		<constant name="FORMAT_R16" value="39" enum="Format">
+			OpenGL texture format [code]GL_R16[/code] where there's one component, a 16-bit unsigned normalized integer value.
```
```suggestion
			OpenGL texture format [code]GL_R16[/code] where there's one component, a 16-bit unsigned normalized integer value. Since the value is normalized, each component is clamped between [code]0.0[/code] and [code]1.0[/code] (inclusive).
```

Same for the other notes.
